src/video_core/command_classes/vic.cpp

```diff
@@ -0,0 +1,189 @@
+    u8* state_offset = reinterpret_cast<u8*>(&vic_state) + offset * sizeof(u32);
+    std::memcpy(state_offset, &arguments, sizeof(u32));
+}
+
+void Vic::ProcessMethod(Vic::Method method, const std::vector<u32>& arguments) {
+    LOG_DEBUG(HW_GPU, "Vic method 0x{:X}", static_cast<u32>(method));
+    VicDeviceWrite(static_cast<u32>(method), arguments[0]);
+
+    switch (method) {
+    case Method::Execute:
+        Execute();
+        break;
+    case Method::SetConfigStructOffset:
+        config_struct_address = gpu.MemoryManager().GpuAddressFromPinned(arguments[0]);
+        break;
+    case Method::SetOutputSurfaceLumaOffset:
+        output_surface_luma_address = gpu.MemoryManager().GpuAddressFromPinned(arguments[0]);
+        break;
+    case Method::SetOutputSurfaceChromaUOffset:
+        output_surface_chroma_u_address = gpu.MemoryManager().GpuAddressFromPinned(arguments[0]);
+        break;
+    case Method::SetOutputSurfaceChromaVOffset:
+        output_surface_chroma_v_address = gpu.MemoryManager().GpuAddressFromPinned(arguments[0]);
+        break;
+    default:
+        LOG_DEBUG(Service_NVDRV, "NV-Vic unimplemented method {}", static_cast<u32>(method));
+    }
+}
+
+void Vic::Execute() {
+    if (output_surface_luma_address == 0) {
+        // TODO(ameerj): Invetigate Link's Awakening bug causing this.
+        LOG_CRITICAL(Service_NVDRV, "VIC LUMA ADDRESS NOT SET. RECEIVED PINNED 0x{:X}",
+                     vic_state.output_surface.luma_offset);
+        // This seems to be the fallback actual address to write to.
+        output_surface_luma_address = gpu.MemoryManager().GpuAddressFromPinned(
+            vic_state.surfacex_slots[0][0].luma_offset - 1);
+        return;
+    }
+
+    const VicConfig config{gpu.MemoryManager().Read<u64>(config_struct_address + 0x20)};
+
+    switch (static_cast<VideoPixelFormat>(config.pixel_format.Value())) {
+    case VideoPixelFormat::Rgba8: {
+        LOG_DEBUG(Service_NVDRV, "Writing RGB Frame");
+        const auto frame = nvdec_processor->GetFrame();
+
+        if (!frame || frame->width == 0 || frame->height == 0) {
+            return;
+        }
+
+        if (scaler_ctx == nullptr || frame->width != scaler_width ||
+            frame->height != scaler_height) {
+            sws_freeContext(scaler_ctx);
+            scaler_ctx = nullptr;
+            // FFmpeg returns all frames in YUV420, convert it into RGBA format
+            scaler_ctx =
+                sws_getContext(frame->width, frame->height, AV_PIX_FMT_YUV420P, frame->width,
+                               frame->height, AV_PIX_FMT_RGBA, 0, nullptr, nullptr, nullptr);
+
+            scaler_width = frame->width;
+            scaler_height = frame->height;
+        }
+
+        // Get RGB frame
+        const std::size_t linear_size = frame->width * frame->height * 4;
+        u8* rgb_frame_buffer = static_cast<u8*>(av_malloc(linear_size));
```

Comment:
```suggestion
        using AVMallocPtr = std::unique_ptr<u8, decltype(&av_free)>;
        AVMallocPtr rgb_frame_buffer{static_cast<u8*>(av_malloc(linear_size)), av_free};
```

This way the code won't leak if any other exit paths are added in the future.